File: src/Python/pipeline_services/examples/benchmark01_pipeline_run.py
from __future__ import annotations

import time
from typing import Dict, List

from pipeline_services.core.pipeline import Pipeline
from pipeline_services.examples.text_steps import append_marker, strip, to_lower


def main() -> None:
    pipeline = Pipeline("benchmark01_pipeline_run", True)
    pipeline.add_action(strip)
    pipeline.add_action(to_lower)
    pipeline.add_action(append_marker)

    input_value = "  Hello Benchmark  "
    warmup_iterations = 1000
    iterations = 10_000

    warmup_index = 0
    while warmup_index < warmup_iterations:
        pipeline.run(input_value)
        warmup_index += 1

    total_pipeline_nanos = 0
    action_totals: Dict[str, int] = {}
    action_counts: Dict[str, int] = {}
    action_names: List[str] = []

    start_ns = time.perf_counter_ns()
    iteration_index = 0
    while iteration_index < iterations:
        result = pipeline.execute(input_value)
        total_pipeline_nanos += result.total_nanos

        for timing in result.timings:
            if timing.action_name in action_totals:
                action_totals[timing.action_name] += timing.elapsed_nanos
                action_counts[timing.action_name] += 1
            else:
                action_totals[timing.action_name] = timing.elapsed_nanos
                action_counts[timing.action_name] = 1
                action_names.append(timing.action_name)

        iteration_index += 1
    end_ns = time.perf_counter_ns()
    wall_nanos = end_ns - start_ns

    print("iterations=", iterations)
    print("wallMs=", float(wall_nanos) / 1_000_000.0)
    print("avgPipelineUs=", float(total_pipeline_nanos) / float(iterations) / 1_000.0)
    print("avgActionUs=")
    for action_name in action_names:
        nanos_total = action_totals[action_name]
        count_total = action_counts[action_name]
        print("  ", action_name, "=", float(nanos_total) / float(count_total) / 1_000.0)


if __name__ == "__main__":
    main()

File: src/typescript/src/pipeline_services/examples/benchmark01_pipeline_run.ts
import { Pipeline, now_ns } from "../../index.js";
import { append_marker, strip, to_lower } from "./text_steps.js";

async function main(): Promise<void> {
  const pipeline = new Pipeline<string>("benchmark01_pipeline_run", true)
    .addAction(strip)
    .addAction(to_lower)
    .addAction(append_marker);

  const inputValue = "  Hello Benchmark  ";
  const warmupIterations = 1_000;
  const iterations = 10_000;

  for (let index = 0; index < warmupIterations; index += 1) {
    await pipeline.run(inputValue);
  }

  let totalPipelineNanos = 0n;
  const actionTotals = new Map<string, bigint>();
  const actionCounts = new Map<string, number>();
  const actionNames: Array<string> = [];

  const startNanos = now_ns();
  for (let index = 0; index < iterations; index += 1) {
    const result = await pipeline.runDetailed(inputValue);
    totalPipelineNanos += result.totalNanos;

    for (const timing of result.actionTimings) {
      if (actionTotals.has(timing.actionName)) {
        actionTotals.set(
          timing.actionName,
          actionTotals.get(timing.actionName)! + timing.elapsedNanos,
        );
        actionCounts.set(
          timing.actionName,
          (actionCounts.get(timing.actionName) ?? 0) + 1,
        );
      } else {
        actionTotals.set(timing.actionName, timing.elapsedNanos);
        actionCounts.set(timing.actionName, 1);
        actionNames.push(timing.actionName);
      }
    }
  }
  const wallNanos = now_ns() - startNanos;

  // eslint-disable-next-line no-console
  console.log("iterations=", iterations);
  // eslint-disable-next-line no-console
  console.log("wallMs=", Number(wallNanos) / 1_000_000.0);
  // eslint-disable-next-line no-console
  console.log(
    "avgPipelineUs=",
    Number(totalPipelineNanos) / Number(iterations) / 1_000.0,
  );
  // eslint-disable-next-line no-console
  console.log("avgActionUs=");
  for (const actionName of actionNames) {
    const nanosTotal = actionTotals.get(actionName)!;
    const countTotal = actionCounts.get(actionName)!;
    // eslint-disable-next-line no-console
    console.log(
      "  ",
      actionName,
      "=",
      Number(nanosTotal) / Number(countTotal) / 1_000.0,
    );
  }
}

void main();
